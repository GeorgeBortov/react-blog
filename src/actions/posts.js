import database from '../firebase/firebase';

// ADD_POST
export const addPost = (post) => ({
    type: 'ADD_POST',
    post
});

export const startAddPost = (postData = {}) => {
    return (dispatch, getState) => {
        const authorID = getState().auth.uid;
        const {
            title = '',
            body = '',
            createdAt = 0,
            authorName = '',
        } = postData;
        const post = { title, body, createdAt, authorName, authorID }
        
        return database.ref(`posts`).push(post).then((ref) => {
            dispatch(addPost({
                id: ref.key,
                ...post
            }));
        });
    }
};

// REMOVE_EXPENSE
export const removePost = ({ id } = {}) => ({
    type: 'REMOVE_POST',
    id
});

export const startRemovePost = ({ id } = {}) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`posts/${id}`).remove().then(() => {
            dispatch(removePost({ id }));
        });
    }
};

// EDIT_EXPENSE
export const editPost = (id, updates) => ({
    type: 'EDIT_POST',
    id,
    updates
});

export const startEditPost = (id, updates) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`posts/${id}`).update(updates).then(() => {
            dispatch(editPost(id, updates));
        });
    }
};

// SET_POSTS
export const setPosts = (posts) => ({
    type: 'SET_POSTS',
    posts
});

export const startSetPosts = () => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        // return database.ref(`users/${uid}/posts`)
        return database.ref(`posts`)
        .once('value')
        .then((snapshot) => {
            const posts = [];

            snapshot.forEach((childSnapshot) => {
                posts.push({
                    id: childSnapshot.key,
                    ...childSnapshot.val()
                });
            })
            dispatch(setPosts(posts));
        });
    }
};