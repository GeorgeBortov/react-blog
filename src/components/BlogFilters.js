import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';

export class BlogFilters extends React.Component {
    state = {
        
    };
    render() {
        return (
            <div className="page-header">
                <div className="content-container">
                    <form>
                        <div className="input-group">
                            <div className="input-group__item">
                                <input
                                    type="text"
                                    className="text-input"
                                    placeholder="Search posts by title"
                                    // value={this.props.filters.text}
                                    // onChange={this.onTextChange}
                                />
                            </div>
                            <div className="input-group__item">
                                <select
                                    className="select"
                                    // value={this.props.filters.sortBy}
                                    // onChange={this.onSortChange}
                                >
                                    <option value="date">Sort by date</option>
                                    <option value="cur_user">Select only my posts</option>
                                </select>
                            </div>
                        </div>
                    </form>
                    <Link className="button" to="/create">Add Post</Link>
                </div>
            </div>
        )
    }
}



export default connect(undefined, undefined)(BlogFilters);