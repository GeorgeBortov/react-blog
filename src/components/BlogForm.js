import React from 'react';
import { connect } from 'react-redux';
import JoditEditor from "jodit-react";
import moment from 'moment';

export class BlogForm extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            title: props.post ? props.post.title : '',
            body: props.post ? props.post.body : '',
            createdAt: moment(),
            authorName: props.userName ? props.userName : (props.post.authorName ? props.post.authorName : '' ),
            error: ''
        }
    }
    onTitleChange = (e) => {
        const title = e.target.value;
        this.setState(() => ({ title }));
    };
    onAuthorChange = (e) => {
        const authorName = e.target.value;
        this.setState(() => ({ authorName }));
    };
    onBodyChange = (value) => {
        const body = value;
        this.setState(() => ({ body }));
    };
    onSubmit = (e) => {
        e.preventDefault();

        if (!this.state.title || !this.state.body) {
            // 'Please provide description and amount'
            this.setState(() => ({ error: 'Please provide title and post body' }));
        } else {
            this.setState(() => ({ error: '' }));
            this.props.onSubmit({
                title: this.state.title,
                body: this.state.body,
                createdAt: this.state.createdAt.valueOf(),
                authorName: this.state.authorName
            });
        }
    };
    config = {
		readonly: false // all options from https://xdsoft.net/jodit/doc/
	}
    render() {
        return (
            <form className="form" onSubmit={this.onSubmit}>
                {this.state.error && <p className="form__error">{this.state.error}</p>}
                <input
                    type="text"
                    placeholder="Post title"
                    autoFocus
                    className="text-input form--title"
                    value={this.state.title}
                    onChange={this.onTitleChange}
                />
                <JoditEditor
                    editorRef={this.setRef}
                    value={this.state.body}
                    config={this.config}
                    onChange={this.onBodyChange}
                />
                {
                    !this.props.userName &&
                    <input
                        type="text"
                        placeholder="Author"
                        className="text-input form--author"
                        value={this.state.authorName}
                        onChange={this.onAuthorChange}
                    />
                }
                
                <div>
                    <button className="button">Save Post</button>
                </div>
            </form>
        )
    }
}

const mapStateToProps = (state) => ({
    userName: state.auth.name || ''
});

export default connect(mapStateToProps, undefined)(BlogForm);