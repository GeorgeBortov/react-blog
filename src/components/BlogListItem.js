import React from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import moment from 'moment';
import trimText from '../selectors/trim-text';
export const BlogListItem = ({ id, title, trimText, authorName, createdAt, uid, authorID, }) => (
    <div className="list-item">
        <div>
            <h3 className="list-item__title">{title}</h3>
            <div className="list-item__text">
                {trimText}
            </div>
        </div>
        <div className="list-item__data">
            <span className="list-item__author">Posted by {authorName !== '' ? authorName : 'Anonymous'}</span>
            <span className="list-item__time">{moment(createdAt).format('MMMM Do YYYY')} at {moment(createdAt).format('h:mm')}</span>
        </div>
        <div className="list-item__buttons">
            <Link className="button" to={`/read/${id}`}>Read</Link>
            {   
                uid === authorID &&
                    <Link className="button" to={`/edit/${id}`}>Edit</Link>
                
            }
            
        </div>
    </div>
    
);

const mapStateToProps = (state, ownProps) => ({
    uid: state.auth.uid,
    trimText: trimText(ownProps.body, 500)
});

const ConnectedBlogListItem = connect(mapStateToProps)(BlogListItem);

export default ConnectedBlogListItem;