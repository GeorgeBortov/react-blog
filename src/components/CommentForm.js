import React from 'react';
import { connect } from 'react-redux';
import JoditEditor from "jodit-react";
import moment from 'moment';

export class BlogForm extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            rate: props.post ? props.post.body : '',
            body: props.post ? props.post.body : '',
            createdAt: moment(),
            authorName: props.userName,
            error: ''
        }
        this.baseState = this.state
    }
    handleRateChange = (e) => {
        const rate = e.target.value;
        this.setState(() => ({ rate })); 
    };
    onAuthorChange = (e) => {
        const authorName = e.target.value;
        this.setState(() => ({ authorName }));
    };
    onBodyChange = (value) => {
        const body = value;
        this.setState(() => ({ body }));
    };
    onSubmit = (e) => {
        e.preventDefault();

        if (!this.state.authorName || !this.state.body || !this.state.rate) {
            // 'Please provide description and amount'
            this.setState(() => ({ error: 'Please fill in all fields and rate the post' }));
        } else {
            e.target.reset();
            this.props.onSubmit({
                rate: this.state.rate,
                body: this.state.body,
                createdAt: this.state.createdAt.valueOf(),
                authorName: this.state.authorName
            });
            this.setState(this.baseState);    
        }
    };
    config = {
		readonly: false // all options from https://xdsoft.net/jodit/doc/
	}
    render() {
        return (
            <div className="comment-form">
                
                <h4>Write a commpent</h4>
                
                <form className="form" onSubmit={this.onSubmit}>
                    {this.state.error && <p className="form__error">{this.state.error}</p>}
                    <span className="rating">
                        <input type="radio" id="star5" name="rating" value="5" onBlur={this.handleRateChange}  /><label htmlFor="star5" title="Rocks!">5 stars</label>
                        <input type="radio" id="star4" name="rating" value="4" onBlur={this.handleRateChange}  /><label htmlFor="star4" title="Pretty good">4 stars</label>
                        <input type="radio" id="star3" name="rating" value="3" onBlur={this.handleRateChange}  /><label htmlFor="star3" title="Meh">3 stars</label>
                        <input type="radio" id="star2" name="rating" value="2" onBlur={this.handleRateChange}  /><label htmlFor="star2" title="Kinda bad">2 stars</label>
                        <input type="radio" id="star1" name="rating" value="1" onBlur={this.handleRateChange}  /><label htmlFor="star1" title="Sucks big time">1 star</label>
                        <span className="txt">Please rate:</span>
                    </span>
                    {
                        !this.props.userName &&
                        <input
                            type="text"
                            placeholder="Your name"
                            className="text-input form--author"
                            value={this.state.authorName}
                            onChange={this.onAuthorChange}
                        />
                    }
                    <JoditEditor
                        editorRef={this.setRef}
                        value={this.state.body}
                        config={this.config}
                        onChange={this.onBodyChange}
                    /> 
                    <div>
                        <button className="button">Publish</button>
                    </div>
                </form>
            </div>
        )
    }
}

const mapStateToProps = (state) => ({
    userName: state.auth.name || state.auth.email
});

export default connect(mapStateToProps, undefined)(BlogForm);