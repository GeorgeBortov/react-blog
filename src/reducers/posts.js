// Posts Reducer

const postsReducerDefaultState = {
    postsArr: [],
    comentsArr: [],
    postTotalNumb: 0
};

export default (state = postsReducerDefaultState, action) => {
    switch (action.type) {
        case 'ADD_POST':
            return {
                ...state,
                postsArr: [
                    ...state.postsArr,
                    action.post
                ]
            };
        case 'ADD_COMMENT':
            return {
                ...state,
                comentsArr: [
                    ...state.comentsArr,
                    action.comment
                ]
                // postsArr: state.postsArr.comments.concat(action.comment)
                
            };
        case 'SET_COMMENTS':
            return {
                ...state,
                comentsArr: action.comments, 
            };
        case 'REMOVE_POST':
            return {
                ...state,
                postsArr: state.postsArr.filter(({ id }) => id !== action.id)
            };
        case 'EDIT_POST':
            return {
                ...state,
                postsArr: state.postsArr.map((post) => {
                    if (post.id === action.id) {
                        return {
                            ...post,
                            ...action.updates
                        }
                    } else {
                        return post;
                    }
                })
            };
        case 'SET_POSTS':
            return {
                ...state,
                postsArr: action.posts,
                postTotalNumb: action.postsNumb
            };
        default:
            return state;
    }
};
