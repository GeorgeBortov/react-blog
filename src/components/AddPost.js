import React from 'react';
import { connect } from 'react-redux';
import BlogForm from './BlogForm';
import { startAddPost } from '../actions/posts';

export class AddPost extends React.Component {
    onSubmit = (post) => {
        this.props.startAddPost(post);
        this.props.history.push('/');
    }
    render() {
        return (
            <div>
                <div className="page-header">
                    <div className="content-container">
                        <h1 className="page-header__title">Add post</h1>
                    </div>
                </div>
                <div className="content-container">
                    <BlogForm
                        onSubmit={this.onSubmit}
                    />
                </div>
            </div>
        );
    }
}

const mapDispatchToProps = (dispatch) => ({
    startAddPost: (post) => dispatch(startAddPost(post))
});

export default connect(undefined, mapDispatchToProps)(AddPost);