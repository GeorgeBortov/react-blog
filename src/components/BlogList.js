import React from 'react';
import { connect } from 'react-redux';
import BlogListItem from './BlogListItem';


export const BlogList = (props) => {
    return (
        <div className="content-container">
            <div className="list-body">
                {   
                    props.posts.length === 0 ? (
                        <div className="list-item list-item--message">
                            <span>No posts</span>
                        </div>
                    ) : (
                        props.posts.map((post) => {
                            return <BlogListItem key={post.id} {...post} />
                        })
                    )
                }
            </div>
        </div>
    )
};

const mapStateToProps = (state) => ({
    posts: state.posts
});

const ConnectedBlogList = connect(mapStateToProps)(BlogList);

export default ConnectedBlogList;