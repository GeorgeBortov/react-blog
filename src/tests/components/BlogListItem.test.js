import React from 'react';
import { shallow } from 'enzyme';
import posts from '../fixtures/posts';
import { BlogListItem } from '../../components/BlogListItem';
import trimText from '../../selectors/trim-text';

test('should render BlogListItem with post and disabled editable link because uid is undefined', () => {
    const wrapper = shallow(<BlogListItem 
        {...posts[0]}
        trimText={trimText(posts[0].body, 255)}
    />);
    expect(wrapper).toMatchSnapshot();
});

test('should add editable link to post', () => {
    const uid = 'thisismytestuid';
    const wrapper = shallow(<BlogListItem 
        {...posts[0]}
        uid={uid}
    />);
    expect(wrapper).toMatchSnapshot();
});
