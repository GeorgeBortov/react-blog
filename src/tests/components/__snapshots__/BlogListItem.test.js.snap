// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should add editable link to post 1`] = `
<div
  className="list-item"
>
  <div>
    <h3
      className="list-item__title"
    >
      test title 1
    </h3>
    <div
      className="list-item__text"
    />
  </div>
  <div
    className="list-item__data"
  >
    <span
      className="list-item__author"
    >
      Posted by 
      George Bortov
    </span>
    <span
      className="list-item__time"
    >
      January 1st 1970
       at 
      1:00
    </span>
  </div>
  <div
    className="list-item__buttons"
  >
    <Link
      className="button"
      replace={false}
      to="/read/1"
    >
      Read
    </Link>
    <Link
      className="button"
      replace={false}
      to="/edit/1"
    >
      Edit
    </Link>
  </div>
</div>
`;

exports[`should render BlogListItem with post and disabled editable link because uid is undefined 1`] = `
<div
  className="list-item"
>
  <div>
    <h3
      className="list-item__title"
    >
      test title 1
    </h3>
    <div
      className="list-item__text"
    >
      Lorem ipsum
 dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Dui saute irure dolor...
    </div>
  </div>
  <div
    className="list-item__data"
  >
    <span
      className="list-item__author"
    >
      Posted by 
      George Bortov
    </span>
    <span
      className="list-item__time"
    >
      January 1st 1970
       at 
      1:00
    </span>
  </div>
  <div
    className="list-item__buttons"
  >
    <Link
      className="button"
      replace={false}
      to="/read/1"
    >
      Read
    </Link>
    <span
      className="button button--disabled"
    >
      Edit
    </span>
  </div>
</div>
`;
